README.md
# ruby-installation
Painless Ruby Installation (Autoinstaller Ruby)

![Progress-1](https://github.com/zeroc0d3/ruby-installation/blob/master/snapshot/install_ruby1.png)
![Progress-2](https://github.com/zeroc0d3/ruby-installation/blob/master/snapshot/install_ruby2.png)

## Dashboard

$VERSION = The Latest Version Stable
```
==========================================================================
  __________                  _________ _______       .___________        
  \____    /___________  ____ \_   ___ \   _  \    __| _/\_____  \  LAB  
    /     // __ \_  __ \/  _ \/    \  \//  /_\  \  / __ |   _(__  <       
   /     /\  ___/|  | \(  <_> )     \___\  \_/   \/ /_/ |  /       \      
  /_______ \___  >__|   \____/ \______  /\_____  /\____ | /______  /      
          \/   \/                     \/       \/      \/        \/       
--------------------------------------------------------------------------
# RUBY INSTALLATION SCRIPT :: ver-$VERSION
==========================================================================
# BEGIN PROCESS..... (Please Wait)  
# Start at: 2017-11-19 19:49:17  

[ 2017-11-19 19:49:17 ] ##### Install Ruby Version: 
[ 2017-11-19 19:49:17 ]       2.4.2 

[ 2017-11-19 19:49:17 ] ##### Using Ruby Package: 
[ 2017-11-19 19:49:17 ]       rbenv 

[ 2017-11-19 19:49:17 ] ##### Archiving Old Ruby Packages: 

[ 2017-11-19 19:49:17 ] :: [ ✔ ]  Old Ruby Packages Archived... 

[ 2017-11-19 19:49:17 ] ##### Download RBENV Repository: 

[ 2017-11-19 19:49:17 ] :: [ ✔ ]  Ruby Installed... 

[ 2017-11-19 19:49:17 ] ##### Load Environment: 
[ 2017-11-19 19:49:17 ]       /home/zeroc0d3/.zshrc 

[ 2017-11-19 19:49:17 ] :: [ ✔ ]  Environment Loaded... 

[ 2017-11-19 19:49:17 ] ##### Ruby Version: 
[ 2017-11-19 19:49:17 ]       ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-linux] 

[ 2017-11-19 19:49:17 ] ##### Path Ruby: 
[ 2017-11-19 19:49:17 ]       /home/zeroc0d3/.rbenv/shims/ruby   

[ 2017-11-19 19:49:17 ] ##### Path Gem: 
[ 2017-11-19 19:49:17 ]       /home/zeroc0d3/.rbenv/shims/gem   

[ 2017-11-19 19:49:17 ] ##### Install Bundle: 
[ 2017-11-19 19:49:17 ]       /home/zeroc0d3/.rbenv/shims/gem install bundle   

[ 2017-11-19 19:49:17 ] :: [ ✔ ]  Bundle Installed... 
==========================================================================
# Finish at: 2017-11-19 19:49:17  
# END PROCESS.....  
```

## Features
- [X] Using `rbenv` Package Manager
- [X] Using `rvm` Package Manager
- [X] Support Docker Installation

## Documentation
* Configuration & How-To, see
[**Wiki Documentation**](https://github.com/zeroc0d3/ruby-installation/wiki)

## Roadmap
* - [X] Check target Ruby version installation
* - [X] Check target Ruby package manager (rbenv / rvm)
* - [X] Check old Ruby Installation, if exist then backup / archive it
* - [X] Running Installation with all configuration (version & package manager) 
* - [X] Load Environment after successfully installation
* - [X] Validate Installation

## License
[**MIT License**](https://github.com/zeroc0d3/ruby-installation/blob/master/LICENSE)

CHANGELOG
## version 1.7.0
====================
* Fixing initialize packages
* Fixing default packages Installer (rbenv)

## version 1.6.0
====================
* Fixing rvm installation path
* Set global archive for rvm & rbenv
* Change initialize default package ruby (rbenv / rvm) with "0" (rbenv) and "1" (rvm)

## version 1.5.4
====================
* Change environment script from `#!/bin/sh` to `#!/usr/bin/env sh`

## version 1.5.3
====================
* Change message color
* Change identity select shell to "VAR_SHELL"
* Fixing check version (${RUBY_VERSION} from Docker environment)
* Fixing check ruby package (${RUBY_PACKAGE} from Docker environment)

## version 1.5.2
====================
* Add title script

## version 1.5.1
====================
* Fixing path for loading environment in ruby installation (install_ruby.sh)
* Fixing path for loading environment & reload shell in gems installation (gems.sh)

## version 1.5
====================
* Remove "docker" folder in this repository. You can clone / forked with other repository:
  1. Docker-Lab (https://github.com/zeroc0d3/docker-lab)
  2. CentOS-Ruby (https://github.com/zeroc0d3lab/centos-ruby)
* Fixing path for loading environment in gems installation (gems.sh)

## version 1.4
====================
* New version of gems.sh file to install ruby dependency packages (bundle install)

## version 1.3
====================
* Update message running installation
* Using ruby.sh for temporary load environment configuration
* Fixing reload shell after success installation

## version 1.2
====================
* Fixing environment bash / zsh reload shell
* Dynamic USERNAME & PATH_HOME

## version 1.1
====================
* Fixing Environment & Reload Shell

## version 1.0
====================
* Add reload shell for bash & zsh

## version 0.9.1
====================
* Installer with rbenv as default package manager

## version 0.9
====================
* Installer with rvm as default package manager


gems.sh
#!/usr/bin/env sh

TITLE="BUNDLE INSTALLATION SCRIPT"   # script name
VER="1.5.3"                          # script version
USERNAME=`echo $USER`                # default username
PATH_HOME=`echo $HOME`               # default home path 

### Path Installation Ruby Package Manager ###
RBENV_ROOT="$PATH_HOME/.rbenv"
RVM_ROOT="/usr/local/rvm"
WHAT_SHELL=`$SHELL -c 'echo $0'`

get_time() {
  DATE=`date '+%Y-%m-%d %H:%M:%S'`
}

logo() {
  clear
  echo "\033[22;32m==========================================================================\033[0m"
  echo "\033[22;31m  __________                  _________ _______       .___________        \033[0m"
  echo "\033[22;31m  \____    /___________  ____ \_   ___ \\   _  \    __| _/\_____  \  LAB  \033[0m"
  echo "\033[22;31m    /     // __ \_  __ \/  _ \/    \  \//  /_\  \  / __ |   _(__  <       \033[0m"
  echo "\033[22;31m   /     /\  ___/|  | \(  <_> )     \___\  \_/   \/ /_/ |  /       \      \033[0m"
  echo "\033[22;31m  /_______ \___  >__|   \____/ \______  /\_____  /\____ | /______  /      \033[0m"
  echo "\033[22;31m          \/   \/                     \/       \/      \/        \/       \033[0m"
  echo "\033[22;32m--------------------------------------------------------------------------\033[0m"
  echo "\033[22;32m# $TITLE :: ver-$VER                                                      \033[0m"
}

header() {
  logo
  echo "\033[22;32m==========================================================================\033[0m"
  get_time
  echo "\033[22;37m# BEGIN PROCESS..... (Please Wait)  \033[0m"
  echo "\033[22;33m# Start at: $DATE  \033[0m\n"
}

footer() {
  echo "\033[22;32m==========================================================================\033[0m"
  get_time
  echo "\033[22;33m# Finish at: $DATE  \033[0m"
  echo "\033[22;37m# END PROCESS.....  \033[0m\n" 
}

load_env() {
  get_time
  echo "\033[22;34m[ $DATE ] ##### Load Environment: \033[0m" 

  ### Running Ruby Environment ###
  PWD=`pwd` 
  cd $PWD; /bin/sh ./ruby.sh
    
  if [ -d "$RBENV_ROOT" ] 
  then
    if [ "$WHAT_SHELL" = "`which zsh`" ] || [ "$WHAT_SHELL" = "zsh" ]
    then  
      echo "\033[22;32m[ $DATE ]       $PATH_HOME/.zshrc \033[0m\n"
    else
      if [ "$WHAT_SHELL" = "`which bash`" ] || [ "$WHAT_SHELL" = "bash" ]
      then
        echo "\033[22;32m[ $DATE ]       $PATH_HOME/.bashrc \033[0m\n"
      else 
        echo "\033[22;32m[ $DATE ]       $PATH_HOME/.bashrc \033[0m\n"
      fi
    fi
  else
    echo "\033[22;32m[ $DATE ]       $RVM_ROOT/scripts/rvm \033[0m\n"
  fi    
  get_time
  echo "\033[22;32m[ $DATE ] :: [ ✔ ] \033[22;32m Environment Loaded... \033[0m\n"
}

validate_installation() {
  get_time
  echo "\033[22;34m[ $DATE ] ##### Ruby Version: \033[0m" 
  RUBY=`which ruby`
  RUBY_V=`$RUBY -v`
  echo "\033[22;32m[ $DATE ]       $RUBY_V \033[0m\n"
  get_time
  echo "\033[22;34m[ $DATE ] ##### Path Ruby: \033[0m" 
  echo "\033[22;32m[ $DATE ]       $RUBY   \033[0m\n"  
  get_time
  echo "\033[22;34m[ $DATE ] ##### Path Gem: \033[0m" 
  GEM=`which gem`
  echo "\033[22;32m[ $DATE ]       $GEM   \033[0m\n"
}

install_bundle() {
  get_time
  echo "\033[22;34m[ $DATE ] ##### Install Bundle: \033[0m" 
  echo "\033[22;32m[ $DATE ]       $GEM install bundle   \033[0m\n"
  $GEM install bundle
  echo ""
  get_time
  echo "\033[22;32m[ $DATE ] :: [ ✔ ] \033[22;32m Bundle Installed... \033[0m\n"
}

reload_shell() {
  if [ -d "$RBENV_ROOT" ] 
  then
    exec $SHELL
  else
    source $RVM_ROOT/scripts/rvm
  fi
}

install_package() {
  get_time
  echo "\033[22;34m[ $DATE ] ##### Bundle Installation Package: \033[0m" 
  BUNDLE=`which bundle`
  echo "\033[22;32m[ $DATE ]       $BUNDLE install             \033[0m\n"
  $BUNDLE install
  echo ""
  get_time
  echo "\033[22;32m[ $DATE ] :: [ ✔ ] \033[22;32m Package Installed... \033[0m"
}

main() {
  header
  load_env
  validate_installation
  install_bundle
  install_package
  footer
}

### START HERE ###
main

dummy.sh
#!/usr/bin/env sh

TITLE="RUBY INSTALLATION SCRIPT"     # script name
VER="1.5.4-dummy"                    # script version
DEFAULT_VERSION="2.4.2"              # default version installation 
DEFAULT_PACKAGE="rbenv"              # default package installation
INSTALL_VERSION=$DEFAULT_VERSION
INSTALL_PACKAGE=$DEFAULT_PACKAGE
USERNAME=`echo $USER`                # default username
PATH_HOME=`echo $HOME`               # default home path 

### Path Installation Ruby Package Manager ###
RBENV_ROOT="$PATH_HOME/.rbenv"
RVM_ROOT="/usr/local/rvm"

### Path Backup Old Ruby Package Manager ###
SNAP_BACKUP=`date '+%Y%m%d%H%M'`
PATH_BACKUP_RBENV=$PATH_HOME"/__rbenv_"$SNAP_BACKUP
PATH_BACKUP_RVM=$PATH_HOME"/__rvm_"$SNAP_BACKUP
VAR_SHELL=`$SHELL -c 'echo $0'`
  
get_time() {
  DATE=`date '+%Y-%m-%d %H:%M:%S'`
}

logo() {
  clear
  echo "\033[22;32m==========================================================================\033[0m"
  echo "\033[22;31m  __________                  _________ _______       .___________        \033[0m"
  echo "\033[22;31m  \____    /___________  ____ \_   ___ \\   _  \    __| _/\_____  \  LAB  \033[0m"
  echo "\033[22;31m    /     // __ \_  __ \/  _ \/    \  \//  /_\  \  / __ |   _(__  <       \033[0m"
  echo "\033[22;31m   /     /\  ___/|  | \(  <_> )     \___\  \_/   \/ /_/ |  /       \      \033[0m"
  echo "\033[22;31m  /_______ \___  >__|   \____/ \______  /\_____  /\____ | /______  /      \033[0m"
  echo "\033[22;31m          \/   \/                     \/       \/      \/        \/       \033[0m"
  echo "\033[22;32m--------------------------------------------------------------------------\033[0m"
  echo "\033[22;32m# $TITLE :: ver-$VER                                                      \033[0m"
}

header() {
  logo
  echo "\033[22;32m==========================================================================\033[0m"
  get_time
  echo "\033[22;37m# BEGIN PROCESS..... (Please Wait)  \033[0m"
  echo "\033[22;33m# Start at: $DATE  \033[0m\n"
}

footer() {
  echo "\033[22;32m==========================================================================\033[0m"
  get_time
  echo "\033[22;33m# Finish at: $DATE  \033[0m"
  echo "\033[22;37m# END PROCESS.....  \033[0m\n" 
}

check_version() {
  ## RUBY_VERSION from Docker Environment
  if [ "${RUBY_VERSION}" != "" ]
  then
    INSTALL_VERSION=${RUBY_VERSION} 
  fi
}

check_ruby_package() {
  ## RUBY_PACKAGE from Docker Environment
  if [ "${RUBY_PACKAGE}" != "" ]
  then
    INSTALL_PACKAGE=${RUBY_PACKAGE}   
  fi
}

cleanup() {
  get_time
  echo "\033[22;34m[ $DATE ] ##### Archiving Old Ruby Packages: \033[0m" 

  # if [ "$INSTALL_PACKAGE" = "rbenv" ]
  # then
  #   if [ -d "$RBENV_ROOT" ]; then 
  #     mv $RBENV_ROOT $PATH_BACKUP_RBENV 
  #   fi 
  # else 
  #   if [ -d "$RVM_ROOT" ]; then 
  #     sudo mv $RVM_ROOT $PATH_BACKUP_RVM 
  #   fi     
  # fi
  # sudo rm -f /etc/profile.d/rvm.sh
  echo ""
  get_time
  echo "\033[22;32m[ $DATE ] :: [ ✔ ] \033[22;32m Old Ruby Packages Archived... \033[0m\n"
}

load_env() {
  get_time
  echo "\033[22;34m[ $DATE ] ##### Load Environment: \033[0m" 

  ### Running Ruby Environment ###
  PWD=`pwd` 
  cd $PWD; /bin/sh ./ruby.sh
    
  if [ "$INSTALL_PACKAGE" = "rbenv" ]
  then
    if [ "$VAR_SHELL" = "`which zsh`" ] || [ "$VAR_SHELL" = "zsh" ]
    then  
      echo "\033[22;32m[ $DATE ]       $PATH_HOME/.zshrc \033[0m\n"
    else
      if [ "$VAR_SHELL" = "`which bash`" ] || [ "$VAR_SHELL" = "bash" ]
      then
        echo "\033[22;32m[ $DATE ]       $PATH_HOME/.bashrc \033[0m\n"
      else 
        echo "\033[22;32m[ $DATE ]       $PATH_HOME/.bashrc \033[0m\n"
      fi
    fi
  else
    echo "\033[22;32m[ $DATE ]       $RVM_ROOT/scripts/rvm \033[0m\n"
  fi    
  get_time
  echo "\033[22;32m[ $DATE ] :: [ ✔ ] \033[22;32m Environment Loaded... \033[0m\n"
}

reload_shell() {
  if [ "$INSTALL_PACKAGE" = "rbenv" ]
  then
    exec $SHELL
  else
    source $RVM_ROOT/scripts/rvm
  fi
}

install_ruby() {
  get_time
  echo "\033[22;34m[ $DATE ] ##### Install Ruby Version: \033[0m" 
  echo "\033[22;32m[ $DATE ]       $INSTALL_VERSION \033[0m\n"
  get_time
  echo "\033[22;34m[ $DATE ] ##### Using Ruby Package: \033[0m" 
  echo "\033[22;32m[ $DATE ]       $INSTALL_PACKAGE \033[0m\n"
  cleanup
  if [ "$INSTALL_PACKAGE" = "rbenv" ]
  then
    #-----------------------------------------------------------------------------
    # Get repo rbenv
    #-----------------------------------------------------------------------------
    get_time
    echo "\033[22;34m[ $DATE ] ##### Download RBENV Repository: \033[0m" 

    # git clone https://github.com/rbenv/rbenv.git $RBENV_ROOT \
    #   && git clone https://github.com/rbenv/ruby-build.git $RBENV_ROOT/plugins/ruby-build 
    
    # #-----------------------------------------------------------------------------
    # # Install Ruby with rbenv (default)
    # #-----------------------------------------------------------------------------
    # $RBENV_ROOT/bin/rbenv install $INSTALL_VERSION \
    #   && $RBENV_ROOT/bin/rbenv global $INSTALL_VERSION \
    #   && $RBENV_ROOT/bin/rbenv rehash 
    
    echo ""
    get_time
    echo "\033[22;32m[ $DATE ] :: [ ✔ ] \033[22;32m Ruby Installed... \033[0m\n"
  else
    #-----------------------------------------------------------------------------
    # Get repo rvm
    #-----------------------------------------------------------------------------
    get_time
    echo "\033[22;34m[ $DATE ] ##### Download RVM Repository: \033[0m" 
  
    # curl -sSL https://rvm.io/mpapis.asc | gpg2 --import \
    #   && curl -sSL https://raw.githubusercontent.com/wayneeseguin/rvm/master/binscripts/rvm-installer | sudo bash -s stable \
    #   && sudo usermod -a -G rvm root \
    #   && sudo usermod -a -G rvm $USERNAME \
    #   && sudo cp ruby.sh /etc/profile.d/rvm.sh \
    #   && umask g+w \
    #   && sudo $RVM_ROOT/bin/rvm get head --auto-dotfiles 

    # #-----------------------------------------------------------------------------
    # # Install Ruby with rvm (alternatives)
    # #-----------------------------------------------------------------------------
    # $RVM_ROOT/bin/rvm install $INSTALL_VERSION \
    #   && $RVM_ROOT/bin/rvm use $INSTALL_VERSION --default 
  
    echo ""
    get_time
    echo "\033[22;32m[ $DATE ] :: [ ✔ ] \033[22;32m Ruby Installed... \033[0m\n"
  fi
}

validate_installation() {
  get_time
  echo "\033[22;34m[ $DATE ] ##### Ruby Version: \033[0m" 
  RUBY=`which ruby`
  RUBY_V=`$RUBY -v`
  echo "\033[22;32m[ $DATE ]       $RUBY_V \033[0m\n"
  get_time
  echo "\033[22;34m[ $DATE ] ##### Path Ruby: \033[0m" 
  echo "\033[22;32m[ $DATE ]       $RUBY   \033[0m\n"  
  get_time
  echo "\033[22;34m[ $DATE ] ##### Path Gem: \033[0m" 
  GEM=`which gem`
  echo "\033[22;32m[ $DATE ]       $GEM   \033[0m\n"
}

install_bundle() {
  get_time
  echo "\033[22;34m[ $DATE ] ##### Install Bundle: \033[0m" 
  echo "\033[22;32m[ $DATE ]       $GEM install bundle   \033[0m\n"
  # $GEM install bundle
  echo ""
  get_time
  echo "\033[22;32m[ $DATE ] :: [ ✔ ] \033[22;32m Bundle Installed... \033[0m"
}

main() {
  header
  check_version
  check_ruby_package
  install_ruby
  load_env
  validate_installation
  install_bundle
  footer
  reload_shell
}

### START HERE ###
main


install_ruby.sh
#!/usr/bin/env sh

TITLE="RUBY INSTALLATION SCRIPT"     # script name
VER="1.7.0"                          # script version
DEFAULT_VERSION="2.4.2"              # default version installation
DEFAULT_PACKAGE="0"                  # default package installation (0 = rbenv, 1 = rvm)
DEFAULT_PACKAGE_NAME="rbenv"
INSTALL_VERSION=$DEFAULT_VERSION
INSTALL_PACKAGE=$DEFAULT_PACKAGE
USERNAME=`echo $USER`                # default username
PATH_HOME=`echo $HOME`               # default home path

### Path Installation Ruby Package Manager ###
RBENV_ROOT="$PATH_HOME/.rbenv"
RVM_ROOT="/usr/local/rvm"

### Path Backup Old Ruby Package Manager ###
SNAP_BACKUP=`date '+%Y%m%d%H%M'`
PATH_BACKUP_RBENV=$PATH_HOME"/__rbenv_"$SNAP_BACKUP
PATH_BACKUP_RVM=$PATH_HOME"/__rvm_"$SNAP_BACKUP
VAR_SHELL=`$SHELL -c 'echo $0'`

get_time() {
  DATE=`date '+%Y-%m-%d %H:%M:%S'`
}

logo() {
  clear
  echo "\033[22;32m==========================================================================\033[0m"
  echo "\033[22;31m  __________                  _________ _______       .___________        \033[0m"
  echo "\033[22;31m  \____    /___________  ____ \_   ___ \\   _  \    __| _/\_____  \  LAB  \033[0m"
  echo "\033[22;31m    /     // __ \_  __ \/  _ \/    \  \//  /_\  \  / __ |   _(__  <       \033[0m"
  echo "\033[22;31m   /     /\  ___/|  | \(  <_> )     \___\  \_/   \/ /_/ |  /       \      \033[0m"
  echo "\033[22;31m  /_______ \___  >__|   \____/ \______  /\_____  /\____ | /______  /      \033[0m"
  echo "\033[22;31m          \/   \/                     \/       \/      \/        \/       \033[0m"
  echo "\033[22;32m--------------------------------------------------------------------------\033[0m"
  echo "\033[22;32m# $TITLE :: ver-$VER                                                      \033[0m"
}

header() {
  logo
  echo "\033[22;32m==========================================================================\033[0m"
  get_time
  echo "\033[22;37m# BEGIN PROCESS..... (Please Wait)  \033[0m"
  echo "\033[22;33m# Start at: $DATE  \033[0m\n"
}

footer() {
  echo "\033[22;32m==========================================================================\033[0m"
  get_time
  echo "\033[22;33m# Finish at: $DATE  \033[0m"
  echo "\033[22;37m# END PROCESS.....  \033[0m\n"
}

check_version() {
  ## RUBY_VERSION from Docker Environment
  if [ "${RUBY_VERSION}" != "" ]
  then
    INSTALL_VERSION=${RUBY_VERSION}
  fi
}

check_ruby_package() {
  ## RUBY_PACKAGE from Docker Environment
  if [ "${RUBY_PACKAGE}" != "" ]
  then
    INSTALL_PACKAGE=${RUBY_PACKAGE}
    if [ "$INSTALL_PACKAGE" = "0" ]
    then
      DEFAULT_PACKAGE_NAME="rbenv"
    else
      DEFAULT_PACKAGE_NAME="rvm"
    fi
  fi
}

cleanup() {
  get_time
  echo "\033[22;34m[ $DATE ] ##### Archiving Old Ruby Packages: \033[0m"
  if [ -d "$RBENV_ROOT" ]; then
    mv $RBENV_ROOT $PATH_BACKUP_RBENV
  fi

  if [ -d "$RVM_ROOT" ]; then
    sudo mv $RVM_ROOT $PATH_BACKUP_RVM
    sudo mkdir -p /user/local/rvm
    sudo chown $USER:$USER -R /usr/local/rvm
    sudo rm -f /etc/profile.d/rvm.sh
  fi
  echo ""
  get_time
  echo "\033[22;32m[ $DATE ] :: [ ✔ ] \033[22;32m Old Ruby Packages Archived... \033[0m\n"
}

load_env() {
  get_time
  echo "\033[22;34m[ $DATE ] ##### Load Environment: \033[0m"

  ### Running Ruby Environment ###
  PWD=`pwd`
  cd $PWD; /bin/sh ./ruby.sh

  if [ "$INSTALL_PACKAGE" = "0" ]     ## default (0 = rbenv)
  then
    if [ "$VAR_SHELL" = "`which zsh`" ] || [ "$VAR_SHELL" = "zsh" ]
    then
      echo "\033[22;32m[ $DATE ]       $PATH_HOME/.zshrc \033[0m\n"
    else
      if [ "$VAR_SHELL" = "`which bash`" ] || [ "$VAR_SHELL" = "bash" ]
      then
        echo "\033[22;32m[ $DATE ]       $PATH_HOME/.bashrc \033[0m\n"
      else
        echo "\033[22;32m[ $DATE ]       $PATH_HOME/.bashrc \033[0m\n"
      fi
    fi
  else
    echo "\033[22;32m[ $DATE ]       $RVM_ROOT/scripts/rvm \033[0m\n"
  fi
  get_time
  echo "\033[22;32m[ $DATE ] :: [ ✔ ] \033[22;32m Environment Loaded... \033[0m\n"
}

reload_shell() {
  if [ "$INSTALL_PACKAGE" = "0" ]
  then
    exec $SHELL
  else
    source $RVM_ROOT/scripts/rvm
  fi
}

install_ruby() {
  get_time
  echo "\033[22;34m[ $DATE ] ##### Install Ruby Version: \033[0m"
  echo "\033[22;32m[ $DATE ]       $INSTALL_VERSION \033[0m\n"
  get_time
  echo "\033[22;34m[ $DATE ] ##### Using Ruby Package: \033[0m"
  echo "\033[22;32m[ $DATE ]       $DEFAULT_PACKAGE_NAME \033[0m\n"
  cleanup
  if [ "$INSTALL_PACKAGE" = "0" ]
  then
    #-----------------------------------------------------------------------------
    # Get repo rbenv
    #-----------------------------------------------------------------------------
    get_time
    echo "\033[22;34m[ $DATE ] ##### Download RBENV Repository: \033[0m"

    git clone https://github.com/rbenv/rbenv.git $RBENV_ROOT \
      && git clone https://github.com/rbenv/ruby-build.git $RBENV_ROOT/plugins/ruby-build

    #-----------------------------------------------------------------------------
    # Install Ruby with rbenv (default)
    #-----------------------------------------------------------------------------
    $RBENV_ROOT/bin/rbenv install $INSTALL_VERSION \
      && $RBENV_ROOT/bin/rbenv global $INSTALL_VERSION \
      && $RBENV_ROOT/bin/rbenv rehash

    echo ""
    get_time
    echo "\033[22;32m[ $DATE ] :: [ ✔ ] \033[22;32m Ruby Installed... \033[0m\n"
  else
    #-----------------------------------------------------------------------------
    # Get repo rvm
    #-----------------------------------------------------------------------------
    get_time
    echo "\033[22;34m[ $DATE ] ##### Download RVM Repository: \033[0m"

    curl -sSL https://rvm.io/mpapis.asc | gpg2 --import \
      && curl -sSL https://raw.githubusercontent.com/wayneeseguin/rvm/master/binscripts/rvm-installer | sudo bash -s stable \
      && sudo usermod -a -G rvm root \
      && sudo usermod -a -G rvm $USERNAME \
      && sudo cp ruby.sh /etc/profile.d/rvm.sh \
      && umask g+w \
      && sudo $RVM_ROOT/bin/rvm get head --auto-dotfiles

    #-----------------------------------------------------------------------------
    # Install Ruby with rvm (alternatives)
    #-----------------------------------------------------------------------------
    $RVM_ROOT/bin/rvm install $INSTALL_VERSION \
      && $RVM_ROOT/bin/rvm use $INSTALL_VERSION --default

    echo ""
    get_time
    echo "\033[22;32m[ $DATE ] :: [ ✔ ] \033[22;32m Ruby Installed... \033[0m\n"
  fi
}

validate_installation() {
  get_time
  echo "\033[22;34m[ $DATE ] ##### Ruby Version: \033[0m"
  RUBY=`which ruby`
  RUBY_V=`$RUBY -v`
  echo "\033[22;32m[ $DATE ]       $RUBY_V \033[0m\n"
  get_time
  echo "\033[22;34m[ $DATE ] ##### Path Ruby: \033[0m"
  echo "\033[22;32m[ $DATE ]       $RUBY   \033[0m\n"
  get_time
  echo "\033[22;34m[ $DATE ] ##### Path Gem: \033[0m"
  GEM=`which gem`
  echo "\033[22;32m[ $DATE ]       $GEM   \033[0m\n"
}

install_bundle() {
  get_time
  echo "\033[22;34m[ $DATE ] ##### Install Bundle: \033[0m"
  echo "\033[22;32m[ $DATE ]       $GEM install bundle   \033[0m\n"
  $GEM install bundle
  echo ""
  get_time
  echo "\033[22;32m[ $DATE ] :: [ ✔ ] \033[22;32m Bundle Installed... \033[0m"
}

main() {
  header
  check_version
  check_ruby_package
  install_ruby
  load_env
  validate_installation
  install_bundle
  footer
  reload_shell
}

### START HERE ###
main
